pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -448,13 +552,33 @@ func (plugin *kubenetNetworkPlugin) teardown(namespace string, name string, id k
 			errList = append(errList, err)
 		}
 	}
+
+	iplist, exists := plugin.podIPs[id]
+	if !exists || 0 == iplist.Len() {
+		klog.V(5).Infof("container %s (%s/%s) does not have recorded. ignoring teardown call", id, name, namespace)
+		return nil
+	}
+
+	for _, ip := range iplist.UnsortedList() {
+		klog.V(5).Infof("Removing pod IP %s from shaper for (%s/%s)", ip, name, namespace)
+		// shaper wants /32
+		isV6 := netutils.IsIPv6String(ip)
+		mask := "32"
+		if isV6 {
+			mask = "128"
+		}
+
+		if err := plugin.shaper().Reset(fmt.Sprintf("%s/%s", ip, mask)); err != nil {
+			// Possible bandwidth shaping wasn't enabled for this pod anyways
+			klog.V(4).Infof("Failed to remove pod IP %s from shaper: %v", ip, err)
+		}
+
+		plugin.removePodIP(id, ip)
+	}
 	return utilerrors.NewAggregate(errList)
 }
 
 func (plugin *kubenetNetworkPlugin) TearDownPod(namespace string, name string, id kubecontainer.ContainerID) error {
-	plugin.mu.Lock()
-	defer plugin.mu.Unlock()
```

Comment:
why is the lock grabbing getting removed?

Reply:
it is granular at `removePodIP` and `addPodIP` funcs..

Reply:
thanks!

Reply:
@khenidak as also sprinkled into comments in other parts, just locking in removePodIP and addPodIP isn't enough, there are a lot of other access of plugin.podIPs that are now unprotected.